src/System.Linq/tests/ToArrayToListToDictionaryTests.cs

```diff
@@ -0,0 +1,469 @@
+        public void ToArray_AlwaysCreateACopy()
+        {
+            int[] sourceArray = new int[] { 1, 2, 3, 4, 5 };
+            int[] resultArray = sourceArray.ToArray();
+
+            Assert.NotSame(sourceArray, resultArray);
+            Assert.Equal(sourceArray, resultArray);
+        }
+
+
+        private void RunToArrayOnAllCollectionTypes<T>(T[] items, Action<T[]> validation)
+        {
+            validation(Enumerable.ToArray(items));
+            validation(Enumerable.ToArray(new List<T>(items)));
+            validation(new EnumerableCollectionTest<T>(items).ToArray());
+            validation(new ReadOnlyCollectionTest<T>(items).ToArray());
+            validation(new CollectionTest<T>(items).ToArray());
+        }
+
+
+        [Fact]
+        public void ToArray_WorkWithEmptyCollection()
+        {
+            RunToArrayOnAllCollectionTypes(new int[0],
+                resultArray =>
+                {
+                    Assert.NotNull(resultArray);
+                    Assert.Equal(0, resultArray.Length);
+                });
+        }
+
+        [Fact]
+        public void ToArray_ProduceCorrectArray()
+        {
+            int[] sourceArray = new int[] { 1, 2, 3, 4, 5, 6, 7 };
+            RunToArrayOnAllCollectionTypes(sourceArray,
+                resultArray =>
+                {
+                    Assert.Equal(sourceArray.Length, resultArray.Length);
+                    Assert.Equal(sourceArray, resultArray);
+                });
+
+            string[] sourceStringArray = new string[] { "1", "2", "3", "4", "5", "6", "7", "8" };
+            RunToArrayOnAllCollectionTypes(sourceStringArray,
+                resultStringArray =>
+                {
+                    Assert.Equal(sourceStringArray.Length, resultStringArray.Length);
+                    for (int i = 0; i < sourceStringArray.Length; i++)
+                        Assert.Same(sourceStringArray[i], resultStringArray[i]);
+                });
+        }
+
+
+        [Fact]
+        public void ToArray_TouchCountWithICollection()
+        {
+            CollectionTest<int> source = new CollectionTest<int>(new int[] { 1, 2, 3, 4 });
+            var resultArray = source.ToArray();
+
+            Assert.Equal(source, resultArray);
+            Assert.Equal(1, source.CountTouched);
+        }
+
+
+        [Fact]
+        public void ToArray_ThrowArgumentNullExceptionWhenSourceIsNull()
+        {
+            int[] source = null;
+            Assert.Throws<ArgumentNullException>(() => source.ToArray());
+        }
+
+
+        // Later this behaviour can be changed
+        [Fact]
+        public void ToArray_UseCopyToWithICollection()
+        {
+            CollectionTest<int> source = new CollectionTest<int>(new int[] { 1, 2, 3, 4 });
+            var resultArray = source.ToArray();
+
+            Assert.Equal(source, resultArray);
+            Assert.Equal(1, source.CopyToTouched);
+        }
+
+
+        // Currently not passed
+        //[Fact]
```

Comment:
I'd suggest leaving [Fact] here but adding [ActiveIssue] with the appropriate issue number, probably 1561.